Enable relu fusion with prepacked linear/conv.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35705 Enable relu fusion with prepacked linear/conv.**

Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

kimishpatel committed:
Enable relu fusion with prepacked linear/conv.

Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -81,8 +81,105 @@ void insertPrePackedConv2dOp(std::shared_ptr<Graph>& graph) {
+        %r = prepacked::conv2d_clamp_run(%input, %packed_weight_bias)
+        return (%r) )";
+
+  if (isReluFusable(graph)) {
```
looks like this is checking for all `aten::relu` in the graph but the pattern is based on individual aten::relu?

I think when the input of relu is used once in the pattern it is only going to match the case when the input is used once as well, have you tried removing this check?

kimishpatel replied:
> looks like this is checking for all `aten::relu` in the graph but the pattern is based on individual aten::relu?

Good point. Yes this checks for all relus.
> 
> I think when the input of relu is used once in the pattern it is only going to match the case when the input is used once as well, have you tried removing this check?

Oh I see. I need to remove this and add a test where output of conv is used elsewhere as well and check.

jerryzh168 replied:
btw, this is only true if the value is not input: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/ir/subgraph_matcher.cpp.html#100

kimishpatel replied:
Ah. Makes sense. Actually, I dont need to test. This makes sense. I was only trying to check that relu is not fused with previous node, if the output of the previous node is used elsewhere. If input to relu is input to the pattern graph then there is nothing relu can fuse into.

kimishpatel committed:
Update on "Enable relu fusion with prepacked linear/conv."


Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

[ghstack-poisoned]

dreiss commented on test/test_xnnpack_integration.py:
```diff
@@ -479,24 +479,52 @@ def __init__(self):
+        pattern_count_map["prepacked::conv2d_clamp_prepack"] = -1
+        pattern_count_map["prepacked::linear_clamp_prepack"] = -1
+        pattern_count_map["aten::relu"] = -1
+        validate_transformed_module(M(), pattern_count_map, data_shape, True, True)
```
See https://our.intern.facebook.com/intern/wiki/Facebook-code-smells/boolean-arguments/ .  I think using booleans is fine here, but pass the args by keyword instead of position so it's clear what they mean.

dreiss commented on torch/csrc/jit/passes/graph_rewrite_helper.h:
```diff
@@ -17,6 +18,27 @@ c10::optional<IValue> getIValue(
+  static PatternInfo parse_from_str(std::string pattern_string) {
+    PatternInfo rv{
+        std::move(pattern_string),
+        std::make_unique<Graph>(), decltype(vmap){}};
```
formatting looks off.

dreiss commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -86,6 +86,82 @@ void insertPrePackedConv2dOp(std::shared_ptr<Graph>& graph) {
+  std::string linear_prepack_run_relu_fused = R"(
+    graph(%input, %weight, %bias, %output_min_max):
+        %output_min: float = prim::Constant[value=0.0]()
+        %output_max: float = prim::Constant[value=1.0]()
```
Shouldn't output_max be None for relu?

kimishpatel replied:
Thanks for the catch.

dreiss commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -86,6 +86,82 @@ void insertPrePackedConv2dOp(std::shared_ptr<Graph>& graph) {
+  SubgraphRewriter rewriter;
+
+  std::string linear_prepack_run_relu_fused = R"(
+    graph(%input, %weight, %bias, %output_min_max):
```
What is output_min_max?  Isn't this supposed to appear in the pattern below?

kimishpatel replied:
Graph fuser will complain if the graph being replaced with has a different signature. Thats why I had to keep it here.
 also realized, I think we will need to do constant pooling after this pass, because for every graph being replaced we are introducing the constant of '0'.

dreiss commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -86,6 +86,82 @@ void insertPrePackedConv2dOp(std::shared_ptr<Graph>& graph) {
+        return (%res))";
+
+  rewriter.RegisterRewritePattern(linear_prepack_run_relu,
+      linear_prepack_run_relu_fused);
```
I'm confused.  Above, you said  "if the input of the relu was not to be overwritten then it should not be fused.", but isn't it being fused here?

kimishpatel replied:
It is a stale comment. I removed some code and forgot to remove the comment. Thanks for catching it. Will remove the comment.

AshkanAliabadi commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -488,6 +488,11 @@ void initJITBindings(PyObject* module) {
+      .def(
+          "_jit_pass_fuse_relu_w_prepacked_linear_conv",
+          [](script::Module& module) {
+            return fusePrePackedLinearConvWithRelu(module);
```
Do we also support different variations of ReLU6 and Hardtanh in this rule?

kimishpatel replied:
I did not add that as part of this, but given the patters are just string. We can replace relu with hardtanh. So should be doable.

AshkanAliabadi replied:
Can you please enable ReLU6 as well (if not already part of this PR.)  Reason being that it is used in MobileNetV2 and I am sure it won't be long before we receive a usecase in which that is causing a performance issue.

kimishpatel replied:
Sure. Will do this in this PR itself. Note that ReLU6 in pytorch just maps to hardtanh.

kimishpatel committed:
Update on "Enable relu fusion with prepacked linear/conv."


Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable relu fusion with prepacked linear/conv."


Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

[ghstack-poisoned]

dreiss commented on test/test_xnnpack_integration.py:
```diff
@@ -199,10 +199,10 @@ def forward(self, x):
-               dilations[0] * (kernels[0] - 1) + 1)
-        assume(width + 2 * paddings[1] >=
-               dilations[1] * (kernels[1] - 1) + 1)
+        assume(height + 2 * paddings[0]
+               >= dilations[0] * (kernels[0] - 1) + 1)
+        assume(width + 2 * paddings[1]
+               >= dilations[1] * (kernels[1] - 1) + 1)
```
Please avoid mixing unrelated formatting changes.

kimishpatel replied:
Sorry about this. I just ran autopep for formatting.

dreiss commented on test/test_xnnpack_integration.py:
```diff
@@ -366,29 +366,30 @@ def forward(self, x):
                 scripted_model._c = torch._C._freeze_module(scripted_model._c)
+            if fuse_clamping_ops:
+                torch._C._jit_pass_fuse_clamp_w_prepacked_linear_conv(scripted_model._c)
+            if (prepack_removal):
```
Don't need parens.

dreiss commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -86,6 +87,147 @@ void insertPrePackedConv2dOp(std::shared_ptr<Graph>& graph) {
+  std::string linear_prepack_run_relu_inplace = R"(
+    graph(%input, %weight, %bias, %output_min_max):
+        %packed_weight_bias = prepacked::linear_clamp_prepack(
+            %weight, %bias, %output_min_max, %output_min_max)
```
Does this verify that output_min_max is None?  If not (for example, if it's 1), this would change semantics.

kimishpatel replied:
No this does not check for the values, it simply reroutes the values. Good catch. The place it fails is:
```
conv2d
hardtanh(-1, 1)
hardtanh(0, 6)
```
On pattern like the above one, if you transform conv2d to conv2d_clamp_prepack + conv2d_clamp_run, and then run clamping fusion pass on it you will get:
```
conv2d_clamp_prepack (min=-1, max=1)
conv2d_clamp_run
hardtanh(0, 6)
```
If you run the pass again on this. We will get pattern that actually gives you wrong result. I think we can fix this. Thanks for the catch.

dreiss commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -86,6 +87,147 @@ void insertPrePackedConv2dOp(std::shared_ptr<Graph>& graph) {
+            %weight, %bias, %output_min_max, %output_min_max)
+        %linear_res = prepacked::linear_clamp_run(%input, %packed_weight_bias)
+        %res = aten::relu(%linear_res)
+        return (%res))";
```
Does this verify that linear_res is not used elsewhere?  If not, this change is not semantics-preserving.

kimishpatel replied:
I had a check earlier to do precisely this. However Jerry pointed out here, https://github.com/pytorch/pytorch/pull/35705#discussion_r400559498,  that this is already taken care of by subgraph rewriter.

kimishpatel committed:
Update on "Enable relu fusion with prepacked linear/conv."


Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable relu fusion with prepacked linear/conv."


Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable relu fusion with prepacked linear/conv."


Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

[ghstack-poisoned]

dreiss commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -93,15 +93,12 @@ bool bothMinMaxNone(
   const auto& match_vmap = match.values_map;
   // If value map has dummy_min_max in it, then it means
   // the pattern to be replaced was the one containing hardtanh.
```
Comment is stale.

dreiss commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -93,15 +93,12 @@ bool bothMinMaxNone(
+      "Expected to find dummy_min_max Value in the subgraph to be replaced.");
+  auto dummy_min_max =
+    graph_rewrite_helper::getIValue("dummy_min_max", match_vmap, vmap);
+  return (!dummy_min_max || dummy_min_max.value().isNone());
```
Is `!dummy_min_max` still valid?

kimishpatel replied:
Just guarding against the possibility that instead of None Value, getIValue returns c10::nullopt.

kimishpatel committed:
Update on "Enable relu fusion with prepacked linear/conv."


Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable relu fusion with prepacked linear/conv."


Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable relu fusion with prepacked linear/conv."


Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable relu fusion with prepacked linear/conv."


Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable relu fusion with prepacked linear/conv."


Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable relu fusion with prepacked linear/conv."


Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable relu fusion with prepacked linear/conv."


Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

[ghstack-poisoned]
